scandium-core/src/main/java/org/eclipse/californium/scandium/CookieGenerator.java

```diff
@@ -55,10 +79,17 @@
 	 * Return a fresh HMAC algorithm instance.
 	 */
 	private Mac getHMAC() throws NoSuchAlgorithmException, CloneNotSupportedException, InvalidKeyException {
+
 		lock.readLock().lock();
 		try {
-			if (hmac != null && !isKeyExpired()) {
-				return (Mac) hmac.clone();
+			if (lastSecretKey != null && !isKeyExpired()) {
+				if (hmac != null) {
+					return (Mac) hmac.clone();
+				} else {
+					Mac hmac = Mac.getInstance("HmacSHA256");
+					hmac.init(lastSecretKey);
+					return hmac;
+				}
```

Comment:
What is the purpose of this new code ?

If clone is not supported each new hmac generation will result in `this.hmac=null`. 

Is it for avoid to generate hmac in write lock ?

Reply:
In the write lock, the generateSecretKey() is called, in the read lock not.
So this is intended to create a hmac with the lastSecretKey instead of generate a new one.

Reply:
Ok I get it.

What about the idea to have 2 cookie generators implementation and if this one failed because of clone exception we switch to a simple one (without all of this)

(If we stay with one class, getHMAC and generateCookie should not raise CloneNotSupportedException)

Reply:
I thought of that. But the performance penalty of this PR should be academical. And the changes in this PR are very minimal compared to two generators.